File: language_bindings/VisualBasic/Examples/FullDemo/CustomSelectSource.vb
Imports System.Windows.Forms

Public Class CustomSelectSource
    Private sourceSelected As Boolean
    Private sSourceName As String
    Public Function IsSourceSelected() As Boolean
        Return sourceSelected
    End Function
    Public Function GetSourceName() As String
        Return sSourceName
    End Function

    Private Sub CustomSelectSource_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        sourceSelected = False
        Dim SourceArray As Integer = 0
        DTWAINAPI.DTWAIN_EnumSources(SourceArray)
        Dim nCount As Integer = DTWAINAPI.DTWAIN_ArrayGetCount(SourceArray)
        If nCount <= 0 Then
            Close()
        End If

        ' Display the sources
        Dim CurSource As Integer = 0
        For i As Integer = 0 To nCount - 1
            Dim szName As String
            szName = Space$(256)
            DTWAINAPI.DTWAIN_ArrayGetAtLong(SourceArray, i, CurSource)
            DTWAINAPI.DTWAIN_GetSourceProductName(CurSource, szName, 255)
            listSources.Items.Add(szName.ToString())
        Next
        listSources.SelectedIndex = 0
        ' Display Info about sources
        Dim sText As String = nCount.ToString() & " TWAIN Source(s) Available for Selection"
        editSourceInfo.Text = sText
    End Sub

    Private Sub btnSelect_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSelect.Click
        sSourceName = listSources.SelectedItem.ToString()
        sourceSelected = True
        Me.Close()
    End Sub

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        sourceSelected = False
        Me.Close()
    End Sub
End Class

File: demos/C#/Examples/FullDemo_Dynamic/CustomSelectSource.cs
using System;
using System.Text;
using System.Windows.Forms;
using DTWAIN_ARRAY = System.IntPtr;
using DTWAIN_SOURCE = System.IntPtr;

namespace TWAINDemo
{
    public partial class CustomSelectSource : Form
    {
        private bool sourceSelected;
        private string sSourceName;
        public bool IsSourceSelected() { return sourceSelected; }
        public string GetSourceName() { return sSourceName; }

        public CustomSelectSource()
        {
            InitializeComponent();
        }

        private void CustomSelectSource_Load(object sender, EventArgs e)
        {
            sourceSelected = false;
            DTWAIN_ARRAY SourceArray = IntPtr.Zero;
            DTwainDemo.TwainAPI.DTWAIN_EnumSources(ref SourceArray);
            int nCount = DTwainDemo.TwainAPI.DTWAIN_ArrayGetCount(SourceArray);
            if (nCount <= 0)
                Close();

            // Display the sources
            DTWAIN_SOURCE CurSource = IntPtr.Zero;
            for (int i = 0; i < nCount; ++i)
            {
                StringBuilder szName = new StringBuilder(256);
                DTwainDemo.TwainAPI.DTWAIN_ArrayGetSourceAt(SourceArray, i, ref CurSource);
                DTwainDemo.TwainAPI.DTWAIN_GetSourceProductName(CurSource, szName, 255);
                listSources.Items.Add(szName.ToString());
            }
            listSources.SelectedIndex = 0;
            // Display Info about sources
            string sText = nCount.ToString() + " TWAIN Source(s) Available for Selection";
            editSourceInfo.Text = sText;
        }

        private void btnSelect_Click(object sender, EventArgs e)
        {
            sSourceName = listSources.SelectedItem.ToString();
            sourceSelected = true;
        }
    }
}
